kep/tbd-gpu-aware-scheduling/README.md

```diff
@@ -0,0 +1,408 @@
+
+###### How can an operator determine if the feature is in use by workloads?
+N/A
+
+###### How can someone using this feature know that it is working for their instance?
+By checking non-GPU pods are not getting scheduled on GPU nodes.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+N/A
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+No
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+It expects kubernetes scheduler to be running.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+No
+
+###### Will enabling / using this feature result in introducing new API types?
+No
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+No
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+No
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+No
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+No
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+No
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+Taint all GPU nodes at the beginning with a No Schedule taint, this will ensure no jobs will get scheduled on the nodes
+with taint.
+
+```yaml
+kubectl taint nodes node1 gpuNodeKey=GPUNodeValue:NoSchedule
+```
+
+Write a mutating admission controller, to add a toleration to the pod spec of any workload that requests for GPUs, 
+or use the ExtendedResourceToleration Admission controller
+
+```yaml 
+tolerations:
+- key: "gpuNodeKey"
+  operator: "Equal"
+  value: "GPUNodeValue"
+  effect: "NoSchedule"
+``` 
+
+Since the GPU pods would have tolerations they would run on the tainted GPU nodes.
+CPU pods on the other hand will not run on GPU nodes as they won’t have any tolerations. 
+
+The drawback in this approach is that, either pod creator(client) needs to add tolerations or it needs to be added by 
+kubernetes cluster manager through mutating admission controller without client's knowledge.
+Taints are on nodes and tolerations are on pods. Submitter of pods and managers of nodes are two different teams sitting
+in different orgs. So one can not manage both the taints and tolerations. And it will increase the chances of errors.
```

Comment:
I'm not against this PR for strong reasons, but I still struggle to see the benefits over taint+tolerations+mutating webhook because I don't see (yet) how the drawbacks list ed here don't apply to the extra-filter scenarios.
IOW: it seems to me the new filter will hit pretty much the same scenarios.

Reply:
We’ve evaluated the taints and tolerations approach as well. However, from the standpoint of managing additional node configurations, it requires cluster admins to add and maintain specific taints on GPU nodes and tolerations on GPU-using pods. In contrast, the filter plugin handles these configurations automatically, significantly reducing the administrative burden.

Reply:
One can label GPU-nodes with something like NFD and then inject node affinity to Pods with a webhook preventing scheduling to those nodes unless the Pod really uses GPU resources.

Automated, low maintenance. That's what I'd do in order to avoid maintaining a scheduler build. The gotcha is if some gitops tool will not allow using such a webhook. But usually gitops can be configured.

Reply:
+1 to @uniemimu 

Can we have a summary of administrative burden for both approaches?
I'm still lacking enough evidence this approch bring major (and thus worthy) benefits

Reply:
Here is a summary table @ffromani 

| Aspect                               | **NFD + Taints/Tolerations + Webhook**                   | **Scheduler Plugin (Cluster Admin Patch)**                   | **Scheduler Plugin (Merged into Default Scheduler)**        |
|--------------------------------------|----------------------------------------------------------|-------------------------------------------------------------|-------------------------------------------------------------|
| **Setup Effort**                     | Requires deploying NFD DaemonSet and webhook             | Requires cluster admin to patch, rebuild, and maintain custom scheduler binary | No additional setup; integrated into the default scheduler   |
| **Ongoing Maintenance**              | Minor updates for labels and webhook adjustments         | Ongoing maintenance as updates require re-patching and rebuilding | Minimal; handled with main scheduler updates                |
| **Configuration Complexity**         | Moderate; multiple components need coordination          | Moderate,  cluster admin manages custom scheduler changes     | Simplified; integrated within default scheduler configuration |
| **Node-Level Overhead**              | NFD DaemonSet on each node                               | None; custom scheduler binary affects only the main scheduler process | None; directly uses the main scheduler across all nodes       |
| **Error-Prone Areas**                | Coordination across NFD, taints, and webhook             | Potential issues if patching isn’t consistently maintained  | Lowest; handled entirely within the default scheduler profile |

Reply:
Especially your column for a patched scheduler is missing a few details.

For the Setup Effort, you will need to add the burden of defining the scheduler name for every workload, unless you manage to replace the existing default scheduler, which is not typically possible in CSP environments. If you don't do those workload definition changes by hand, add a webhook. If you do it by hand for every workload, you could do manual changes for workload node affinity by hand also in the left column for NFD. So either remove webhook from the left column or add it to the middle column.

For the ongoing maintenance about labels, the GPU vendors tend to have their own set of labels which has been rather stable, but varies between vendors. The dominating market leader has NFD installation as an option in its GPU device plugin helm chart. It's not like it would be an effort to set up and maintain.

The node-level overhead for the patched scheduler version will only be "none" in case you replace the existing default scheduler, which is usually not possible in CSP managed kubernetes environments. So you typically need an extra node, possibly a big one, if the cluster is big. The NFD worker daemonset is small and in the case of those GPU nodes where you didn't want extra Pods, it shouldn't force you to increase your node counts unless 5 millicores and 128MB of memory is too much. On other nodes then, there is that amount of unnecessary resource usage. So the  row there in my opinion would be "typically 0 new nodes required, plausibly 1 extra node" | "typically 1 extra node required" | "0"

From the overall hassle point of view having a default scheduler forcing this filtering for all users by default is of course the simplest, and the right column rightfully shows that.

What the table doesn't show is whether there is widespread desire to have this filtering happen for all non-GPU workloads. I know you want it, but do everybody? That's a pretty big change to the current way of scheduling. Has this been presented in sig-scheduling?

Reply:
> What the table doesn't show is whether there is widespread desire to have this filtering happen for all non-GPU workloads. I know you want it, but do everybody? That's a pretty big change to the current way of scheduling. Has this been presented in sig-scheduling?

+1 again. I can see a way to have this code accepted in kubernetes-sigs, but changing the default profile is a much more impactful change which would require extensive discussion and evaluation.

Will review the rest shortly.

Reply:
I think the whole discussion is moving around "CSP managed kubernetes environments". Why so ? There are companies (specially) the larger companies who manage their own Kubernetes environments / cluster setups. For them it makes sense to use hardware from cloud as IaaS and then install Kubernetes on top of that by themselves. Everyone is not willing to use GKE or EKS or any other "CSP managed kubernetes environments".
Let's talk about the open source Kubernetes and not club it with CSP goals.

Reply:
> What the table doesn't show is whether there is widespread desire to have this filtering happen for all non-GPU workloads. I know you want it, but do everybody? That's a pretty big change to the current way of scheduling. Has this been presented in sig-scheduling?

If that's part of the process before merging a new plugin to this repo then, let's follow the process. We can present this with sig-scheduling, that's a good suggestion. @gkeesh7 @Zeel-Patel Please prepare for that.

Reply:
From where I'm standing, I don't oppose having the plugin here in this repository. I just don't see much benefit myself, since I believe I could do the same with less of a maintenance burden, basically without having to keep building a custom scheduler. But from considering having alternatives for those who really prefer building their own scheduler, I get it.

There's still yet another approach available for achieving the same thing, and that is scheduler extenders. With extenders you wouldn't have to rebuild the whole scheduler, you'd only have to build the extender and hook it up. The filter functionality is available in extenders. The downside of those is that you still need to configure the scheduler, and you still need to maintain your own scheduler extender container. Plus you take a performance hit in scheduling, which can be mitigated by the use of caches and only passing in node names to the extender. But you still take a measurable performance hit.

My view is that real benefits in terms of making things simple start to only surface if you get this to Kubernetes upstream all the way into the default profile where it would be enabled by default, and **that** requires sig-scheduling to buy the idea.